api/internal/plugins/loader/loader.go

```diff
@@ -29,6 +30,10 @@ import (
 type Loader struct {
 	pc *types.PluginConfig
 	rf *resmap.Factory
+
+	// absolutePluginHome caches the location of a valid plugin root directory.
+	// It should only be set once the directory's existence has been confirmed.
+	absolutePluginHome string
```

Comment:
> as to see PluginConfig as part of the pre-build setup, so each attempt to use a plugin could assume pre-validated config, and the errors that might happen during build wouldn't - ideally - be because of bad config

There are still related errors that could happen though, right? E.g. root path exists, but specific plugin referenced does not. Since this actually isn't general plugin requirement but rather config for exec and go plugins specifically, I would argue that it isn't possible to correctly validate it ahead of build, because we don't know whether we need it yet--hence this PR delaying the check until it is relevant.

> it would be nice to be able to assume once the build/inflation process begins, PluginConfig becomes a readonly struct.

With the field moved to the loader, this continues to be true for `PluginConfig`. If I move it back down, it can't both be set ahead of time and not cause the issue this PR attempts to solve... unless we have a field for the potential error as well as one for the path and have the loader look for and report the error if applicable? I think that would work, but it feels less straightforward.

> This PR moves PluginConfig::AbsPluginHome 'up' here, which seems out of place in the scheme of things.

Could it be thought of as plugin loader config specifically? Not all plugins need it, and the loader is the thing that actually needs/uses this information.

Reply:
yes, sg.

Reply:
when no perfect answer presents itself, its good to just sweep up a bit and call it a day.

I like how Go makes it possible to do so.  The way you can make something implement an interface without actually depending on that interface is wonderful for making slow improvement and avoiding getting into the architectural messes common in Java and C++.